Metadata API: Avoid raising securesystemslib exceptions

MVrachev opened the pull request:
Fixes #1646 

**Description of the changes being introduced by the pull request**:

In this pr, I aim to remove the need of our users to import securesystemslib
in order to catch the exceptions thrown by `tuf/metadata/api.py`. 
I either reexport securesystemslib exceptions or
reraise them by using tuf exceptions.

Reexporting securesystemslib exceptions is required so that users
of code that throws these exceptions doesn't have to import
securesystemslib to handle them.

A couple of additional fixes are added to this pr as well.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Fix type annotation in test_api

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Metadata API: add exception tests

Add missing tests testing raising documented
exceptions for "Metadata.sign()",
"Metadata.to_file()" and "Metadata.from_file()".

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -1371,10 +1365,7 @@ def from_data(
+            except (sslib_exceptions.FormatError,) as e:
                 raise exceptions.UnsupportedAlgorithmError(
                     f"Unsupported algorithm '{algorithm}'"
                 ) from e
```
I mentioned this elsewhere already but after looking at this for a while I've convinced myself that we could just handle both of them and return something generic like UnsignedMetadataError like the verify functions do, no? then I think we we can remove UnsupportedAlgorithmError altogether.

jku replied:
ah no, it is still used by TargetFile.from_data()

jku commented:
I think we can remove the last SSLIB error as well https://github.com/theupdateframework/python-tuf/pull/1725#discussion_r780236043

opinions?

joshuagl commented:
> I think we can remove the last SSLIB error as well [#1725 (comment)](https://github.com/theupdateframework/python-tuf/pull/1725#discussion_r780236043)
> 
> opinions?

:+1:
Removing all sslib exceptions from the python-tuf public API seems like a good goal.

MVrachev commented:
@jku and @lukpueh you can have a look into my list commit where I actually get rid of all securesystemslibs exceptions and there is no longer need for reexporting.

lukpueh commented:
> @jku and @lukpueh you can have a look into my list commit where I actually get rid of all securesystemslibs exceptions and there is no longer need for reexporting.

Is this ready for review? Or should I wait until we land #1725?

MVrachev commented:
You can review the last three commits.
Even though https://github.com/theupdateframework/python-tuf/pull/1734/commits/8403a891b8749fde4a7415b440ae7029f7dd6b3d and https://github.com/theupdateframework/python-tuf/pull/1734/commits/9090da2a797509624756e27b4ef6b5681297e3a1 should be squashed.

MVrachev commented:
@jku @lukpueh this pr is ready for review after the merge of #1725.

jku commented on tests/test_api.py:
```diff
@@ -212,6 +223,21 @@ def test_sign_verify(self) -> None:
             targets_key.verify_signature(md_obj)
 
+    def test_sign_failures(self) -> None:
+        # Make sure there is a case where we throw UnsignedMetadataError.
+        # If one day securesystemslib change their implementation and throw
+        # something different than UnsignedMetadataError,
+        # then make sure the metadata API sign doc will change too.
```
this comment is incorrect (sslib does not raise UnsignedMetadataError) and very long

MVrachev replied:
Ohhh... such a bad comment. :D 
Yea, I am definitely updating it.

jku commented:
I think there's a misunderstanding. My comments on securesystemslib.StorageError vs OSError were made in the context of the client: there a securesystemslib error just makes life harder for no good reason. In Metadata API there was a conscious decision to abstract storage (for the repository use case). Removing StorageError completely makes no sense here: we might want to expose StorageError in our exceptions though so that users don't need to import SSLIB explicitly.

I agree about the other SSLIB errors.

Commit:
Reexport sslib StorageError and change docs

Reexport securesystemslib StorageError, so that our users can catch
it without importing securesystemslib.
The securesystemslib StorageError makes sense in the context of
metadata API, because it supports different storage interfaces and
this exception is denoting all possible errors that could arrise
from using any kind of storage interface.

Additionally, I changed the places where we mention that StorageError
is thrown, so that our users will know they can directly import it
from tuf/api/exceptions.py instead of importing securesystemslib.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
sign sslib.exceptions -> UnsignedMetadataError

Catch Metadata.sign() securesystemslib exceptions and instead throw
a more general UnsignedMetadataError exception.
We don't want to expose securesystemslib exceptions and it's better
to replace them with a more general exception that could be easily
handled.

As the signer is an argument implementing securesystemslib.signer.Signer
interface we don't know what exception will it throw.
That's why we need to catch all possible exceptions during signing and
raise UnsignedMetadataError.
That is the same reason why we should move the serialization outside
the "try" block, so a tuf.api.serialization.SerializationError can
propagate and warn the user that 'signed' cannot be serialized.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
@jku the pr is updated.
You are correct we do want to keep the `StorageError` exceptions as we do provide an option to use the `storage_backend` argument and we want to be able to throw exceptions a `StorageError` which generalizes exceptions from different storage interfaces.

jku reviewed (approved):
I agree with this. Would appreciate other maintainer opinions (@lukpueh ?) on my suggestion to re-export StorageError: I feel like that's a good strategy but maybe a little unusual?

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -305,9 +304,7 @@ def sign(
         Raises:
             tuf.api.serialization.SerializationError:
                 'signed' cannot be serialized.
```
This is not raised anymore. Can you move the serialize call out of the try block?

lukpueh replied:
My comment from above is nonsense. What I wanted to say is, if we except all below when calling `sign(serialize...)` and re-raise as `UnsignedMetadataError`, then `SerializationError` won't be raised anymore.

So either we take the call to serialize out of the try block below or we update the docstring

MVrachev replied:
I decided to move the serialization outside the try block so that the serialization error will propagate as I don't think it's good to hide it with `UnsignedMetadataError` exception.

jku replied:
wow I completely missed this. Good solution I think.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -320,7 +317,15 @@ def sign(
+            signature = signer.sign(signed_serializer.serialize(self.signed))
+        except (
+            sslib_exceptions.CryptoError,
+            sslib_exceptions.UnsupportedAlgorithmError,
```
This `except` block assumes that `signer` is an `SSlibSigner`, but the caller could pass any implementation of the `Signer`  interface. Ideally the `Signer` interface would define what exceptions to expect here, like storage or serialization interfaces do (maybe this is worth a ticket in securesystemslib?).

Should we just except all and re-raise as UnsignedMetadataError?

jku replied:
good catch. For this PR (and 1.0) the last suggestion sounds reasonable to me.

MVrachev replied:
Really great catch.
Totatlly forgot that `signed_serializer` is an object implementing the general `securesystemslib.signer.Signer` interface.

MVrachev replied:
> Ideally the Signer interface would define what exceptions to expect here, like storage or serialization interfaces do (maybe this is worth a ticket in securesystemslib?).


The `securesystemslib` issue about this: https://github.com/secure-systems-lab/securesystemslib/issues/386

lukpueh reviewed (commented):
Regarding `StorageError`, I think re-exporting makes sense. See inline for more comments...
